Include/pyport.h

```diff
@@ -160,6 +160,27 @@ typedef int Py_ssize_clean_t;
 #   define PY_FORMAT_SIZE_T "z"
 #endif
 
+
+/* _Py_HAVE_TYPEOF and _Py_TYPEOF can opportunistically support an equivalent
+ * of GCC's typeof extension where possible. It is not possible to get
+ * equivalent behavior on all platforms, so all uses of _Py_TYPEOF should be
+ * guarded by _Py_HAVE_TYPEOF.
+ */
+#if defined(__GNUC__) || defined(__clang__) || defined(__cplusplus)
+#   define _Py_HAVE_TYPEOF 1
+#   if defined(__cplusplus)
+#       define _Py_TYPEOF(x) decltype(x)
+#   else
+#       define _Py_TYPEOF(x) __typeof__(x)
```

Comment:
I am not entirely sure if maybe this should be `__extension__ ( __typeof__(x) )` to allow its use in `-pedantic` mode or something.